File: 022. Generate Parentheses/C++/Solution.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        class Utils {
        public:
            void parenthesis(vector<string> &result, string current, int left, int right) {
                if (left == 0 && right == 0) {
                    result.push_back(current);
                } else if (left >= 0 && right >= left) {
                    parenthesis(result, current + "(", left - 1, right);
                    parenthesis(result, current + ")", left, right - 1);
                }
            }
        };
        vector<string> result;
        Utils utils;
        utils.parenthesis(result, "", n, n);
        return result;
    }
};

File: 022. Generate Parentheses/Python/Solution.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        def parenthesis(result, current, left, right):
            if left == 0 and right == 0:
                result.append(current)
            elif left >= 0 and right >= left:
                parenthesis(result, current + '(', left - 1, right)
                parenthesis(result, current + ')', left, right - 1)

        result = []
        parenthesis(result, '', n, n)
        return result
